Bluetooth: Mesh: Remaining pieces of persistent storage support

jhedberg opened the pull request:
This is a follow-up PR to #7368 

TODO:

- [x] Offload storing/clearing to system workqueue
- [ ] IV Update state handling
- [x] Heartbeat publication
- [x] Model publication
- [x] Model subscription
- [x] Model binding
- [x] Any other Configuration Server model states
- [x] Proper testing

Fixes #7073 once GATT CCC storage has been implemented as part of #7417

Commit:
Bluetooth: Mesh: Redesign element and storage info for models

Keeping the model struct same sized, change the element pointer to two
indexes, and add a flags member that will be used to track pending
storage actions.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Mesh: Convert RPL storage timer into a generic one

To reduce stack consumption, and to avoid blocking the CPU during
network activity, prepare for a generic timer that can be used for
most (possibly all) mesh storage values.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Mesh: Introduce generic storage timer

Instead of having an RPL-specific storage timer, introduce a generic
one that'll eventually be used for all persistent storage.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Mesh: Perform storage clearing also through delayed work

In order not to have a Node Reset consume more stack than other
operations, also perform the related storage writes through the same
delayed work as all other storage updates.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

carlescufi commented on samples/bluetooth/mesh/prj_bbc_microbit.conf:
```diff
@@ -2,8 +2,17 @@ CONFIG_INIT_STACKS=y
+CONFIG_FLASH_PAGE_LAYOUT=y
+CONFIG_FLASH_MAP=y
+CONFIG_FCB=y
+CONFIG_SETTINGS=y
```
Shouldn't this be selected automatically by `CONFIG_BT_SETTINGS`?

jhedberg replied:
Btw, this is a copy from prj.conf that was already merged, so both places will need fixing if I switch to `select`. FWIW, I'm fine with `select` in this case - I was just wondering if we have documented reasoning to guide developers with choosing between `depends on` and `select`.

carlescufi replied:
Now this confuses me. I have been told (can't remember by whom) to always use `select` and not `depends on` so as to avoid the number of lines in a `.conf`. Remember that the difference between Linux and Zephyr is `oldconfig` and using `.config` files directly. In Zephyr `.config` files are treated right now purely as output build files, not as files you keep for your app. I would say that is the reason why `selects` is a better choice for Zephyr.

carlescufi replied:
@SebastianBoe @ulfalizer might want to weigh in

jhedberg replied:
I think the general idea is that you don't end up getting unexpected features into your build, and you don't end up getting conflicts where you're requested a feature but it's trying to select another features who's pre-requisites aren't fulfilled (it might not be available for your board, you don't have some third feature enabled, or something else). I.e. in summary try to eliminate too much implicit/magic stuff happening behind the scenes which you can't see or control very well.

jhedberg replied:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/kbuild/kconfig-language.txt#n128

jhedberg replied:
Btw, based on the above reasoning in kconfig-language.txt, BT_SETTINGS should use `depends on` since SETTINGS has its own dependencies on NFFS || FCB

ulfalizer replied:
Ahh... I see how `depends on` came up now.

One case where you should avoid `select` is for simple dependencies, like below:

```
config FOO_DEP
    bool "foo_dep"

config FOO
    bool "foo"
    select FOO_DEP
```

This is cleaner:

```
config FOO_DEP
    bool "foo_dep"

config FOO
    bool "foo"
    depends on FOO_DEP
```

`select` makes dependencies harder to follow, since you can't tell from looking at a symbol that other things select it. Often you run into dependency issues too. `select` should mostly be restricted to "helper" symbols.

As a bonus, `FOO` will get indented below `FOO_DEP` there if you use `depends on`.

carlescufi replied:
OK, thanks @ulfalizer for the input. Let's go with `select` then.

jhedberg replied:
> OK, thanks @ulfalizer for the input. Let's go with select then.

@carlescufi I think you mean `depends on`? Since SETTINGS is not a "helper" symbol.

carlescufi replied:
oops, yes, I meant `depends on`. Apologies.

carlescufi commented on samples/bluetooth/mesh/prj_bbc_microbit.conf:
```diff
@@ -2,8 +2,17 @@ CONFIG_INIT_STACKS=y
+CONFIG_FLASH_MAP=y
+CONFIG_FCB=y
+CONFIG_SETTINGS=y
+CONFIG_SETTINGS_FCB=y
```
Sorry if I missed this, is it documented somewhere that the sample now uses FCB?

carlescufi replied:
So how does this work then? persistent storage is only enabled when there's no GATT enabled?

jhedberg replied:
The GATT possibility for this sample app is a special-case that you have to explicitly choose with `CONF_FILE=microbit_gatt.conf`. If you don't do that you just get a "normal" mesh app without PB-GATT or GATT Proxy features.

carlescufi replied:
OK thanks for the explanation.

carlescufi commented:
CC @trond-snekvik @thomasstenersen

vikrant8051 commented:
For #BluetoothMesh, is all persistence data storage going to based on FCB ? 
Will it use #NVS:6391, once it merged in future ? 

What is going to be final ? & why ?

jhedberg commented:
> For #BluetoothMesh (or #Bluetooth), is all persistence data storage going to based on FCB?

No, it's based on the settings API, and the settings API currently supports FCB and NFFS as backends. I've enabled FCB in the sample app configuration since I needed to choose one or the other, and FCB is more light-weight.

> Will it use #NVS:6391, once it get merged in future ?
> What is going to be final ? & why ?

Nothing is final, Zephyr is a continuously evolving project :)  NVS wasn't merged (and still isn't) when I started working on this, and persistent storage is something we want for 1.12 for which the merge window closing is coming up very soon. Once we have one working solution, we can look at refactoring it to let NVS be one possible storage option, either directly or as a backend to settings.

jhedberg commented:
Another note in favor of settings over NVS: if there's only a partition and some other subsystem requires a file system, doing settings over NFFS will work as a solution both for settings and that other subsystem, whereas with NVS you'd need to redo the partitioning to have two partitions.

Commit:
Bluetooth: Mesh: Support for storing model bindings persistently

Add support for storing the bound App Keys for each model
persistently. The bindings are stored under the settings key
bt/mesh/s/<mod id>/bind for SIG models and bt/mesh/v/<mod id>/bind for
vendor models.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Mesh: Support for storing model subscriptions persistently

Add support for storing the subscribed group addresses for each model
persistently. The addresses are stored under the settings key
bt/mesh/s/<mod id>/sub for SIG models and bt/mesh/v/<mod id>/sub for
vendor models.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

vikrant8051 commented:
Is Model subscription address get stored?

Because in case of #meshctl, for unprovisioned node I execute 

 discover-unprovisioned on 
 security 0 
 provision xxxx
 menu config
 target 0100 
 appkey-add 1
 bind 0 1 1000
 bind 0 1 1001
 bind 0 1 1002
 bind 0 1 1003
 sub-add 0100 c000 1000
 sub-add 0100 c000 1002
 pub-set 0100 c000 1 0 5 1001
 pub-set 0100 c000 1 0 5 1003

Then reset the board, & execute only

 connect 0000
 menu config
 target 0100 
 pub-set 0100 c000 1 0 5 1001
 pub-set 0100 c000 1 0 5 1003

But my on board LEDs do not react.  So have to execute, 
   sub-add 0100 c000 1000
   sub-add 0100 c000 1002

Commit:
Bluetooth: Mesh: Support for storing model publication persistently

Add support for storing the model publication information
persistently. The addresses are stored under the settings key
bt/mesh/s/<mod id>/pub for SIG models and bt/mesh/v/<mod id>/pub for
vendor models.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
@vikrant8051 please try with the latest version of patches in this PR (just a few minutes old). At least in my own tests restoring both subscription and publication works now.

vikrant8051 commented:
Ok

vikrant8051 commented:
I did but facing same condition...

I executed following command ..

> cd $(zephyr_base)
> 
> git fetch origin pull/7428/head:BluetoothMesh
> 
> git checkout BluetoothMesh
> 
> git rebase BluetoothMesh

jhedberg commented:
@vikrant8051 it really sounds like you might not have the latest patches, since both subscription and publication works for me (I'm using samples/bluetooth/mesh for testing). Can you confirm if you have "Bluetooth: Mesh: Support for storing model publication persistently" as part of your tree?

vikrant8051 commented:
"Bluetooth: Mesh: Support for storing model publication persistently"
Is it another PR ?

vikrant8051 commented:
yes

vikrant8051 commented:
--> git log 

commit 02bceccd86e47d7cca6104874a75e3e3d59a84fa
Author: Johan Hedberg <johan.hedberg@intel.com>
Date:   Wed May 9 21:34:52 2018 +0300

    Bluetooth: Mesh: Support for storing model bindings persistently
    
    Add support for storing the bound App Keys for each model
    persistently. The bindings are stored under the settings key
    bt/mesh/s/<mod id>/bind for SIG models and bt/mesh/v/<mod id>/bind for
    vendor models.
    
    Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

commit 7d545d8121fe755fd8670f019d359244ab8b98b2
Author: Johan Hedberg <johan.hedberg@intel.com>
Date:   Tue May 8 21:18:55 2018 +0300

    Bluetooth: Mesh: Redesign element and storage info for models
    
    Keeping the model struct same sized, change the element pointer to two
    indexes, and add a flags member that will be used to track pending
    storage actions.
    
    Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

commit 2c8e70b30acce6e21b1eb5b4642853fc68d64187
Author: Johan Hedberg <johan.hedberg@intel.com>
Date:   Wed May 9 09:21:54 2018 +0300

    Bluetooth: Mesh: Perform storage clearing also through delayed work
    
    In order not to have a Node Reset consume more stack than other
    operations, also perform the related storage writes through the same
    delayed work as all other storage updates.

vikrant8051 commented:
I think latest things have merged in my local repository. 

But I've observed strange behavior. Now I've to only execute

> connect 0000
> menu config
> target 0100
> sub-add 0100 c000 1000

vikrant8051 commented:
@jhedberg 
Now I directly modified sample/bluetooth/mesh demo example & edit gen_onoff_set_unack() to control LED1 on #nRF52840_PDK board. Let it be Board A.

For provision & configure Board A, I used this commands
discover-unprovisioned on
security 0
provision xxxx
menu config
target 0100
appkey-add 1
bind 0 1 1000
sub-add 0100 c000 1000

For testing, use Generic_OnOff_Client which is executing on another board (Let it be B). If I don't reset Board A then it works as expected but as soon as I reset then have to execute following commands

connect 0000
menu config
target 0100
sub-add 0100 c000 1000

vikrant8051 commented:
Now I realized that publish addresses are getting saved but same is not happening in case of subscription addresses.

jhedberg commented:
@vikrant8051 you're completely right, sorry. I've forgotten to add the appropriate calls to `bt_mesh_store_mod_sub()` to `cfg_srv.c`. I'll update this PR in a moment with the fix.

jhedberg commented:
@vikrant8051 the PR has now been updated and module subscription should be properly stored. Please test.

vikrant8051 commented:
Ok ..

Commit:
Bluetooth: Mesh: Support for storing heartbeat publication persistently

Add support for storing the heartbeat publication persistently. The
information is only stored as "publish indefinitely" or as "periodic
publishing disabled" since we can't know for how long the node is
powered off. The information is stored under the settings key
bt/mesh/HBPub.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Mesh: Support for storing common configuration states

Add support for storing the remaining configuration server model
states (all of which are one byte values). The states are stored under
a single settings key bt/mesh/Cfg.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

vikrant8051 commented:
HI @jhedberg,

This is today observations based on following configuration in case of Node A & Node B,
(Each Node have - 
          
> Gen. OnOFF Server, {LED 1} 
> Gen. OnOFF Client, {Button1 : ON & Button 2: OFF}
> Gen. Level Server,  {LED3... if (Level  < 50) LED3: OFF ...if (Level > 50) LED3: ON }  
> Gen. Level Client, {Button3: publishes Level = 25 .....Button4: publishes Level = 75 }
)

I configured node A with following commands,

> sub-add 0100 c000 1000
> pub-set  0100 c000 1 0 5 1001

> sub-add 0100 c000 1002
> pub-set 0100  c000 1 0 5 1003 

& Node B as

> sub-add 0101 c000 1000
> pub-set  0101 c000 1 0 5 1001

> sub-add 0101 **c001** 1002 
> pub-set 0101 **c001** 1 0 5 1003

---------------------------------------------------------------------------------------------------------------------------------

If I press Button 1/2 on any Node, then LED1 should turn ON/OFF on both Node. 
This is because Generic OnOFF server on both NODEs has same subscription address i.e. 0xC000.

(This is happening as expected)

---------------------------------------------------------------------------------------------------------------------------------

If I press Button 3/4 on any Node, then LED3 should react only on that particular node.
(since NODE A Gen. Level server subscription addr = 0xC000 & 
          NODE B Gen. Level server subscription addr = 0xC001 )

> **But if I pressed Button 3/4 on NODE A, then LED 3 on NODE B are also reacting...which is strange.**


This is not happening in case of NODE B. Here if I pressed  Button 3/4 then LED 3 on only NODE B are reacting. (This is happening as expected)

vikrant8051 commented:
I would like to re-test it to confirm my observations.

vikrant8051 commented:
Yes !! I've confirmed my observation.

Commit:
Bluetooth: Mesh: Fix updating RPL for locally originated messages

We can implicitly trust locally originated messages, so there's no
need to burden the RPL with them.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Mesh: Fix sequence number restoring

The logic for restoring the sequence number was flawed in that it
would not always cause an incremented sequence write upon the
transmission of the first packet. The reason the code didn't work is
that it assumed the stored value was a multiple of SEQ_STORE_RATE,
however since the sequence number is stored in a deferred fashion
that's not always true.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: samples/mesh_demo: Add support for settings-based storage

Enable BT_SETTINGS and remove any custom flash access used for the
sequence number. We do settings_load() before bt_mesh_provision() and
use the return value from the latter to identity that settings_load()
had recoved a full configuration.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

Commit:
Bluetooth: Mesh: Remove sequence number from bt_mesh_provision()

The sequence number was acting as a stop-gap for missing persistent
storage. Now that we have the settings support in place it's no longer
needed.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

jhedberg commented:
@vikrant8051 is what you're describing related to the persistent storage? Could you try the latest version of this PR - there was a fix to locally originated messages and the RPL which might be related.

vikrant8051 commented:
Ok..

vikrant8051 commented:
Now I disable, 
#CONFIG_BT_SETTINGS=y
#CONFIG_FLASH=y
#CONFIG_FLASH_PAGE_LAYOUT=y
#CONFIG_FLASH_MAP=y
#CONFIG_FCB=y
#CONFIG_SETTINGS=y
#CONFIG_SETTINGS_FCB=y

& commented 

/*
	if (IS_ENABLED(CONFIG_SETTINGS)) {
		settings_load();
	}
	*/
Even after that, facing same observation. 

That means it has nothing to do with Persistent storage.

vikrant8051 commented:
@jhedberg 
I've already share my customized version of samples/bluetooth/mesh project (via Google Drive Link) with you which I am using for testing. Could you please quickly check it so that I can use it without any hesitation for testing purpose ?

carlescufi reviewed (approved):
This is too foreign to me for me to review in-depth, but I have read through the code and have not been able to find outstanding issues.

jhedberg commented:
Thanks @carlescufi 
I'll merge this now, and we can then work out any remaining issues or missing features in follow-up PRs.
